File: server/common/protocol.py
import socket
import struct

BUFFER_SIZE = 4096

# Message types (matching Go client)
MESSAGE_TYPE_BET = 1
MESSAGE_TYPE_BATCH = 2
MESSAGE_TYPE_RESPONSE = 3
MESSAGE_TYPE_GET_WINNERS = 4


class BetMessage:
    """Represents a betting request from client"""

    def __init__(self, nombre, apellido, documento, nacimiento, numero):
        self.type = MESSAGE_TYPE_BET
        self.nombre = nombre
        self.apellido = apellido
        self.documento = documento
        self.nacimiento = nacimiento
        self.numero = numero


class BatchMessage:
    """Represents multiple bets sent together"""

    def __init__(self, agency, bets, eof=False):
        self.type = MESSAGE_TYPE_BATCH
        self.agency = agency  # Agency number (1-5)
        self.bets = bets  # List of BetMessage objects
        self.eof = eof

    @classmethod
    def from_data(cls, data):
        """Parse batch message from custom protocol data"""
        if len(data) < 1 or data[0] != MESSAGE_TYPE_BATCH:
            raise ValueError("Invalid batch message")

        content = data[1:].decode("utf-8")
        parts = content.split("|")

        if len(parts) < 3:
            raise ValueError("Invalid batch message format")

        agency = int(parts[0])
        eof = parts[1] == "1"
        bet_count = int(parts[2])

        bets = []
        idx = 3
        for _ in range(bet_count):
            if idx + 4 >= len(parts):
                break
            bet = BetMessage(
                nombre=parts[idx],
                apellido=parts[idx + 1],
                documento=parts[idx + 2],
                nacimiento=parts[idx + 3],
                numero=int(parts[idx + 4]),
            )
            bets.append(bet)
            idx += 5

        return cls(agency, bets, eof)


class GetWinnersMessage:
    """Represents a request to get winners for an agency"""

    def __init__(self, agency):
        self.type = MESSAGE_TYPE_GET_WINNERS
        self.agency = agency

    @classmethod
    def from_data(cls, data):
        """Parse get winners message from custom protocol data"""
        if len(data) < 1 or data[0] != MESSAGE_TYPE_GET_WINNERS:
            raise ValueError("Invalid get winners message")

        content = data[1:].decode("utf-8")
        agency = int(content)
        return cls(agency)


class ResponseMessage:
    """Represents server response to client"""

    def __init__(self, success, error=None, winners=None):
        self.type = MESSAGE_TYPE_RESPONSE
        self.success = success
        self.error = error
        self.winners = winners or []


def read_packet_from(client_socket):
    """Read length-prefixed packet and parse message"""
    # Read length prefix (4 bytes)
    length_data = _read_exact(client_socket, 4)
    length = int.from_bytes(length_data, byteorder='big')

    # Read message data
    data = _read_exact(client_socket, length)

    # Parse based on message type
    if len(data) < 1:
        raise ValueError("Empty message")

    msg_type = data[0]

    if msg_type == MESSAGE_TYPE_BATCH:
        return BatchMessage.from_data(data)
    elif msg_type == MESSAGE_TYPE_GET_WINNERS:
        return GetWinnersMessage.from_data(data)
    else:
        raise ValueError(f"Unknown message type: {msg_type}")


def send_response(client_socket, success, error=None, winners=None):
    """Send response message using custom protocol"""
    # Build response data
    data = bytearray()
    data.append(MESSAGE_TYPE_RESPONSE)

    # Add success flag
    data.extend(("1" if success else "0").encode("utf-8"))
    data.extend(b"|")

    # Add error if present
    if error:
        data.extend(error.encode("utf-8"))
    data.extend(b"|")

    # Add winners if present
    if winners:
        data.extend(str(len(winners)).encode("utf-8"))
        data.extend(b"|")
        for winner in winners:
            data.extend(winner.encode("utf-8"))
            data.extend(b"|")
    else:
        data.extend(b"0|")

    # Send length-prefixed message
    length = len(data)
    length_bytes = length.to_bytes(4, byteorder='big')

    client_socket.send(length_bytes)
    client_socket.send(data)


def _read_exact(sock, n):
    """Read exactly n bytes from socket"""
    data = b""
    while len(data) < n:
        chunk = sock.recv(n - len(data))
        if not chunk:
            raise RuntimeError("Socket connection broken")
        data += chunk
    return data

File: client/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"strconv"
	"strings"
	"syscall"

	"github.com/tinchomorilla/2c25-sistemasDistribudos-tp0-base/client/common"
)

// SerializeMessage converts a message to custom protocol format
func SerializeMessage(message interface{}) ([]byte, error) {
	var data []byte

	switch msg := message.(type) {

	case *BatchMessage:
		data = append(data, byte(MessageTypeBatch))
		content := fmt.Sprintf("%d|%d|%d", msg.Agency, boolToInt(msg.EOF), len(msg.Bets))
		for _, bet := range msg.Bets {
			content += fmt.Sprintf("|%s|%s|%s|%s|%d", bet.Nombre, bet.Apellido, bet.Documento, bet.Nacimiento, bet.Numero)
		}
		data = append(data, []byte(content)...)

	case *GetWinnersMessage:
		data = append(data, byte(MessageTypeGetWinners))
		content := fmt.Sprintf("%d", msg.Agency)
		data = append(data, []byte(content)...)

	default:
		return nil, fmt.Errorf("unsupported message type")
	}

	return data, nil
}

// DeserializeMessage parses custom protocol data into a message
func DeserializeMessage(data []byte) (interface{}, error) {
	if len(data) < 1 {
		return nil, fmt.Errorf("empty message")
	}

	msgType := MessageType(data[0])
	content := string(data[1:])

	switch msgType {
	case MessageTypeResponse:
		return parseResponseMessage(content)
	default:
		return nil, fmt.Errorf("unsupported message type: %d", msgType)
	}
}

// parseResponseMessage parses response message from pipe-delimited content
func parseResponseMessage(content string) (*ResponseMessage, error) {
	parts := strings.Split(content, "|")
	if len(parts) < 3 {
		return nil, fmt.Errorf("invalid response format")
	}

	success := parts[0] == "1"
	errorMsg := parts[1]

	winnerCount, err := strconv.Atoi(parts[2])
	if err != nil {
		return nil, fmt.Errorf("invalid winner count: %w", err)
	}

	var winners []string
	for i := 3; i < 3+winnerCount && i < len(parts); i++ {
		if parts[i] != "" {
			winners = append(winners, parts[i])
		}
	}

	return &ResponseMessage{
		Type:    MessageTypeResponse,
		Success: success,
		Error:   errorMsg,
		Winners: winners,
	}, nil
}

// SendMessage serializes message and sends with length prefix
func SendMessage(conn net.Conn, message interface{}) error {
	data, err := SerializeMessage(message)
	if err != nil {
		return fmt.Errorf("error serializing message: %w", err)
	}

	// Send length prefix
	length := uint32(len(data))
	lengthBytes := make([]byte, common.LENGTH_PREFIX_BYTES)
	binary.BigEndian.PutUint32(lengthBytes, length)

	if err := writeExact(conn, lengthBytes); err != nil {
		if errors.Is(err, syscall.EPIPE) {
			return fmt.Errorf("error sending length: broken pipe")
		}
		return fmt.Errorf("error sending length: %w", err)
	}

	// Send message data
	if err := writeExact(conn, data); err != nil {
		if errors.Is(err, syscall.EPIPE) {
			return fmt.Errorf("error sending data: broken pipe")
		}
		return fmt.Errorf("error sending data: %w", err)
	}

	return nil
}

// RecvMessage reads length-prefixed message and deserializes
func RecvMessage(conn net.Conn, v interface{}) error {
	// Read length prefix (4 bytes)
	lengthBytes := make([]byte, 4)
	if _, err := readExact(conn, lengthBytes); err != nil {
		if err == io.EOF {
			return io.EOF
		}
		return fmt.Errorf("error reading length: %w", err)
	}

	length := binary.BigEndian.Uint32(lengthBytes)

	// Read message data
	data := make([]byte, length)
	if _, err := readExact(conn, data); err != nil {
		if err == io.EOF {
			return io.EOF // Unexpected EOF, incomplete message
		}
		return fmt.Errorf("error reading data: %w", err)
	}

	// Deserialize message
	message, err := DeserializeMessage(data)
	if err != nil {
		return fmt.Errorf("error deserializing message: %w", err)
	}

	// Copy to destination
	switch dst := v.(type) {
	case *ResponseMessage:
		if src, ok := message.(*ResponseMessage); ok {
			*dst = *src
		} else {
			return fmt.Errorf("type mismatch: expected ResponseMessage")
		}
	default:
		return fmt.Errorf("unsupported destination type")
	}

	return nil
}

// readExact reads exactly len(buf) bytes from conn
func readExact(conn net.Conn, buf []byte) (int, error) {
	totalRead := 0
	for totalRead < len(buf) {
		n, err := conn.Read(buf[totalRead:])
		if err != nil {
			return totalRead, err
		}
		totalRead += n
	}
	return totalRead, nil
}

// writeExact writes exactly all bytes in data over conn
func writeExact(conn net.Conn, data []byte) error {
	totalWritten := 0
	for totalWritten < len(data) {
		n, err := conn.Write(data[totalWritten:])
		if err != nil {
			return err
		}
		if n == 0 {
			return fmt.Errorf("connection closed during write")
		}
		totalWritten += n
	}
	return nil
}
